watchtower/wtdb/tower_db.go

```diff
@@ -0,0 +1,577 @@
+func (t *TowerDB) InsertSessionInfo(session *SessionInfo) error {
+	return t.db.Update(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(sessionsBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		sessionBytes := sessions.Get(session.ID[:])
+		if sessionBytes != nil {
+			return ErrSessionAlreadyExists
+		}
+
+		return putSession(sessions, session)
+	})
+}
+
+// InsertStateUpdate stores an update sent by the client after validating that
+// the update is well-formed in the context of other updates sent for the same
+// session. This include verifying that the sequence number is incremented
+// properly and the last applied values echoed by the client are sane.
+func (t *TowerDB) InsertStateUpdate(update *SessionStateUpdate) (uint16, error) {
+	var lastApplied uint16
+	err := t.db.Update(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(sessionsBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		updates := tx.Bucket(updatesBkt)
+		if updates == nil {
+			return ErrUninitializedDB
+		}
+
+		// Fetch the session corresponding to the update's session id.
+		// This will be used to validate that the update's sequence
+		// number and last applied values are sane.
+		session, err := getSession(sessions, update.ID[:])
+		if err != nil {
+			return err
+		}
+
+		// Validate the update against the current state of the session.
+		err = session.AcceptUpdateSequence(
+			update.SeqNum, update.LastApplied,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Validation succeeded, therefore the update is committed and
+		// the session's last applied value is equal to the update's
+		// sequence number.
+		lastApplied = session.LastApplied
+
+		// Store the updated session to persist the updated last applied
+		// values.
+		err = putSession(sessions, session)
+		if err != nil {
+			return err
+		}
+
+		// Finally, create or load the hint bucket for this state
+		// update's hint and write the given update.
+		hints, err := updates.CreateBucketIfNotExists(update.Hint[:])
+		if err != nil {
+			return err
+		}
+
+		var b bytes.Buffer
+		err = update.Encode(&b)
+		if err != nil {
+			return err
+		}
+
+		return hints.Put(update.ID[:], b.Bytes())
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return lastApplied, nil
+}
+
+// DeleteSession removes all data associated with a particular session id from
+// the tower's database.
+func (t *TowerDB) DeleteSession(target SessionID) error {
+	return t.db.Update(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(sessionsBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		updates := tx.Bucket(updatesBkt)
+		if updates == nil {
+			return ErrUninitializedDB
+		}
+
+		// Fail if the session doesn't exit.
+		_, err := getSession(sessions, target[:])
+		if err != nil {
+			return err
+		}
+
+		// Remove the target session.
+		err = sessions.Delete(target[:])
+		if err != nil {
+			return err
+		}
+
+		return updates.ForEach(func(hint, v []byte) error {
```

Comment:
Perhaps we should have an additional index mapping a session to the state hints it's stored? This would allow us to avoid scanning over the _entire database of session updates_ for a delete. Instead, we'd iterate over the index and delete the individual sessions. If we go this route, then we might also need a reference counter somewhere as well.

Reply:
Initially thought we could do a migration that would add this info, since it's possible to do a complete scan and reconstruct these indexes. I went ahead and added them tho and it wasn't too much additional code, lmk what you think f6c9bd6